components/script/dom/dommatrixreadonly.rs

```diff
@@ -966,80 +968,148 @@ pub(crate) fn entries_to_matrix(entries: &[f64]) -> Fallible<(bool, Transform3D<
     }
 }
 
-// https://drafts.fxtf.org/geometry-1/#validate-and-fixup
-pub(crate) fn dommatrixinit_to_matrix(dict: &DOMMatrixInit) -> Fallible<(bool, Transform3D<f64>)> {
-    // Step 1.
-    if dict.parent.a.is_some() &&
-        dict.parent.m11.is_some() &&
-        dict.parent.a.unwrap() != dict.parent.m11.unwrap() ||
-        dict.parent.b.is_some() &&
-            dict.parent.m12.is_some() &&
-            dict.parent.b.unwrap() != dict.parent.m12.unwrap() ||
-        dict.parent.c.is_some() &&
-            dict.parent.m21.is_some() &&
-            dict.parent.c.unwrap() != dict.parent.m21.unwrap() ||
-        dict.parent.d.is_some() &&
-            dict.parent.m22.is_some() &&
-            dict.parent.d.unwrap() != dict.parent.m22.unwrap() ||
-        dict.parent.e.is_some() &&
-            dict.parent.m41.is_some() &&
-            dict.parent.e.unwrap() != dict.parent.m41.unwrap() ||
-        dict.parent.f.is_some() &&
-            dict.parent.m42.is_some() &&
-            dict.parent.f.unwrap() != dict.parent.m42.unwrap() ||
-        dict.is2D.is_some() &&
-            dict.is2D.unwrap() &&
-            (dict.m31 != 0.0 ||
-                dict.m32 != 0.0 ||
-                dict.m13 != 0.0 ||
-                dict.m23 != 0.0 ||
-                dict.m43 != 0.0 ||
-                dict.m14 != 0.0 ||
-                dict.m24 != 0.0 ||
-                dict.m34 != 0.0 ||
-                dict.m33 != 1.0 ||
-                dict.m44 != 1.0)
+/// <https://drafts.fxtf.org/geometry-1/#matrix-validate-and-fixup-2d>
+fn validate_and_fixup_2d(dict: &DOMMatrix2DInit) -> Fallible<[f64; 6]> {
+    // <https://tc39.es/ecma262/#sec-numeric-types-number-sameValueZero>
+    let same_value_zero = |x: f64, y: f64| -> bool { x.is_nan() && y.is_nan() || x == y };
+
+    // Step 1. If if at least one of the following conditions are true for dict,
+    // then throw a TypeError exception and abort these steps.
+    if dict.a.is_some() &&
+        dict.m11.is_some() &&
+        !same_value_zero(dict.a.unwrap(), dict.m11.unwrap()) ||
```

Comment:
non-blocking: it would be nicer to use let chains
```suggestion
    if let Some((a, m11)) = dict.a.zip(dict.m11) &&
        !same_value_zero(a, m11) ||
```

Reply:
Ok

Reply:
Rust compiler doesn't want to do it with error: ``||` operators are not supported in let chain expressions`

Reply:
Right, we could split each check into own if, but it's not worth the hassle IMO, so let's keep it as is.